satpy/tests/test_readers.py

```diff
@@ -751,7 +751,7 @@ def test_no_reader(self):
 
         # without files it's going to be an empty result
         assert group_files([]) == []
-        groups = group_files(self.g16_files)
+        groups = group_files(self.g16_files, reader='abi_l1b')
```

Comment:
@BENR0 can you comment on this?

Reply:
@mraspaud what you are saying makes sense to me. I honestly didn't check what that test does I just made this change based on @djhoese suggestion so maybe he can shed more light on this.

Reply:
Whoa I did not realize what the test was actually trying to accomplish. This is one downside of the generic filename in the reader. It matches everything and has the change to "steal" matches from other readers. Either:

1. This test gets updated to expect 1 group of files (not great).
2. The reader gets updated with group keys to sort by start time...which might work and fix this test.
3. Remove the default file pattern from the reader

Reply:
Well obviously I am against option 3. :-D. I think removing a "feature" because a test fails is not the way to go. I am not sure which of the remaining options is the best way to move forward.

Reply:
Satpy needs a strategy to choose a reader if multiple reader match a file pattern.

Reply:
Reader order may come down to:

https://github.com/pytroll/satpy/blob/c0b69beb9173ded3f4dea13893a7c64b7e5620f2/satpy/readers/__init__.py#L328

in most cases. There may be others. It may get even more complicated when we look at that `glob_configs` function since it is searching not only Satpy's configs but any user configured paths too:

https://github.com/pytroll/satpy/blob/c0b69beb9173ded3f4dea13893a7c64b7e5620f2/satpy/_config.py#L156-L184

Either way, it clearly uses `iglob` which isn't going to be sorted.

As for my option 3 @BENR0, it isn't that the test fails, it is that the failing test represents a use case (as @gerritholl says): which reader gets chosen when a user doesn't specify a reader (they don't know what one they want). With this PR they seem to always get the CF reader even though there is a much better choice.

@gerritholl besides alphabetical order, I'm not sure we want to go down the performance-lossy road of checking all readers for a match against a filename. We *could* implement something that knows that the CF reader and the generic image reader should be last resorts:

```
# pseudo code
generic_choice = None
for reader, reader_info in ...:
    if reader.matches(filename):
        if not reader_info["generic"]:
            return reader
        generic_choice = reader
return generic_choice
```

Reply:
I understand. This discussion reminded me of the conversation regarding overlapping file patterns  in #913 which might be related.

Reply:
Similar, but they are separate bits of code. The overlapping file patterns will likely always be a little difficult. This reader choice one can have a solution, but it might slow some operations down.